### DISCUSSION TITLE
IPC to return a list of open windows in the workspace

### DISCUSSION THREAD

**Discussion by jakesarjeant** [YaLTeR/niri#624] [Ideas]
I love scrollable tiling, but when I have many windows open in a workspace, it can be hard to keep track of which one is where, and I occasionally find myself aimlessly SUPER+H/SUPER+Ling around my workspace looking for the right window.

I think it would be tremendously helpful if my bar could display an ordered list of open windows in the active workspace so that I can see at a glance where the window I'm looking for is. I imagine an IPC message kind of like:

```
$ niri msg layout "DP-1" 1
Column 1:
    blender

Column 2:
    kitty
    kitty

Column 3:
 * firefox
```

So that my bar can display the corresponding icons, e.g.

![Screenshot from 2024-08-24 17-39-15](https://github.com/user-attachments/assets/c3f011bd-5002-4d85-a727-ea507414edf0)

**Comment by YaLTeR** [comment]
Hmm, I'll consider this for #453

**Reply by jetjinser** [reply]
Hi, how does the event stream IPC make this possible? I'm also interested in the features discussed in this discussion: Displays a list of windows of the current workspace on the top bar.

I checked the current [events](https://yalter.github.io/niri/niri_ipc/enum.Event.html) and I have no clue on how to implement it. Any suggestions? Thanks!

**Reply by YaLTeR** [reply]
Yes, though at the moment not ordered. You get a list of workspaces with IDs, and a list of windows which have a workspace_id.

**Reply by jetjinser** [reply]
Ah, got it. Thank you!

But still, I would love a way to make an ordered list of windows possible, I hope it's a "nice to have" feature :)

**Comment by druskus20** [comment]
Yes, would be nice also if moving windows generates events in the EventStream, which currently, I believe, does not. 
I have a use case in mind for a workspaces/windows indicator widget

**Reply by jakesarjeant** [reply]
Yeah, you're right. I'm working on a PR now which will hopefully add events for exactly those kinds of layout changes.

**Reply by druskus20** [reply]
Been thinking about this again - now that we got floating in master! - Actually I for my specific usecase I don't need a list of windows, just a list of columns, ordered by position within the workspace.

**Reply by YaLTeR** [reply]
I think @calops also wanted to look into putting coordinates into IPC, perhaps you could sync with them?

**Reply by calops** [reply]
Indeed I'm working on this, I'll open a PR for this shortly. For now I'm mostly thinking about adding the column index to the windows information (as well as new events), which should cover your use-case if you subscribe to the events stream.

**Reply by kosayoda** [reply]
> Indeed I'm working on this, I'll open a PR for this shortly. For now I'm mostly thinking about adding the column index to the windows information (as well as new events), which should cover your use-case if you subscribe to the events stream.

I would love this feature. Please let me know if I can help.

**Comment by jakesarjeant** [comment]
I've got some time now, so I'll try and put a PR together in the next few days. Shouldn't be that hard since #453 is already merged.

**Reply by jakesarjeant** [reply]
The most straightforward way I can think of to go about this would be to just add an ordered list of open window IDs to `niri::ipc::Workspace`, but that would be a breaking change to the current API, and it would cause a `WorkspacesChanged` event to be fired for every time a window is changed or reordered.

Alternatively (this is what I'm doing now), I could add a new `Layout { workspace_id, columns }` struct and corresponding `LayoutChanged` event. My only concern with this is that it doesn't really follow the "make invalid state irrepresentable" maxim, as it would theoretically be possible to have multiple layouts associated with the same workspace ID.

@YaLTeR do you have any input on this? Maybe you have a better idea since you know your codebase better than I do...

**Reply by YaLTeR** [reply]
FWIW, adding new fields to structs is not a breaking change. But changing the behavior of existing events in such a way that clients written against the previous version become logically incorrect would be a breaking change and we shouldn't do it.

I'm not yet sure how to best expose the rest of the layout over IPC. I don't think I want to do it just now necessarily. I know people also want window positions within a workspace, which I think could be used as a substitute here (you can sort by window X within a workspace). But I don't have a clear design for that either. I think it would make sense to add some kind of `position_within_workspace: Option<(f64, f64)>` to a `Window`, along with a `WindowPositionsChanged { windows: Vec<Window> }` event. This way it's forward-compatible with anything we might want to do in the future (can just set this to `None` if it doesn't make sense).

**Reply by jakesarjeant** [reply]
> FWIW, adding new fields to structs is not a breaking change

It is for any rust users because a) rust requires every field to be mentioned in pattern matching and because b) AFAIK, serde doesn't allow unknown fields by default

> I think it would make sense to add some kind of `position_within_workspace: Option<(f64, f64)>` to a `Window`, along with a `WindowPositionsChanged { windows: Vec<Window> }` event. This way it's forward-compatible with anything we might want to do in the future

Makes sense. I'll go ahead and mock up an implementation of that. Obviously, it's up to you to decide whether you want to merge it and when (I imagine there's some other things in the pipeline that might affect decision making with respect to this); I don't mind running my little fork on my personal setup until then and updating my bar when you decide how to go forward

**Reply by YaLTeR** [reply]
> It is for any rust users

niri-ipc is explicitly *not* rust-semver-compatible for this reason

> serde doesn't allow unknown fields by default

it does allow them

> I imagine there's some other things in the pipeline that might affect decision making with respect to this

Yeah, we'll see

**Reply by jakesarjeant** [reply]
> niri-ipc is explicitly *not* rust-semver-compatible for this reason

Ah, didn't know that

> it does allow them

You're right. It seems my memory has failed me once again...

> Yeah, we'll see

I suppose we will. I'll be watching with interest

**Comment by sommerper** [comment]
@jakesarjeant I currently use this as an alt-tab replacement.

```
#!/bin/bash

# Get window information and format it for fuzzel
window_list=$(niri msg --json windows | jq -r '.[] | "\(.id) | [\(.app_id)] - \(.title)"')

# Present the window list to fuzzel and get the selected window ID
selected=$(echo "$window_list" | fuzzel -d -p "Switch to: " | cut -d'|' -f1)

# If a window was selected, focus it
if [ -n "$selected" ]; then
    niri msg action focus-window --id $selected
fi
```

Maybe you can use that until something better comes along.

**Reply by jakesarjeant** [reply]
Thanks, but I feel like alt-tab style navigation sort of defeats the point of the spacial arrangement made possible by niri — I can navigate quite a lot more intuitively in niri; I just want something I can look to for a quick hint, especially when I'm using an application that I don't use very often

**Comment by wzhchin** [comment]
Same request here.

I am currently trying to adapt a system status monitoring bar for it, and the issue I'm facing is that I can't retrieve the coordinates of each window, which causes the order at the top to be incorrect. Is there any way to obtain the coordinates of all the windows?

The reason for using specific coordinates is that there may be future support for floating windows and sticky windows.

**Bright light theme warning**
If your current environment is not suitable for viewing high-brightness videos, please do not play the video.

https://github.com/user-attachments/assets/afdc55a0-7b81-4f77-806e-381e10ab25d1

**Reply by YaLTeR** [reply]
I'll add it sometime yeah

**Reply by Nitepone** [reply]
Just ran into needing window locations over IPC.

@YaLTeR Is there a detailed proposal for how this should be implemented anywhere? I'm thinking it would be nice to be able to get column & row indices from the "windows" IPC. In addition to true locations.

Would happily pick this up and make a PR for it this week!

**Reply by YaLTeR** [reply]
There's a wip PR at https://github.com/YaLTeR/niri/pull/1265
